api: Improve error handling

peterthomassen opened the pull request:
- better error codes
- make domain DELETE idempotent
- wrap Domain database writes in transaction, commit only when pdns operations are successful

TODO:
- fix testCantDeleteOtherDomains test
    - done
- return a reasonable error if pdns operations fail (currently 500)
    - done (if domain already exists, return 409)
- If a user is acting (GET, DELETE, ...) on a domain that belongs to another user, we return 404 (because the Django queryset is empty). 403 may be more appropriate, but it leaks a bit of information
    - not doing this now, problem is rather esoteric

nils-wisiol commented:
also fix error code for domains that already exist in the pdns db

peterthomassen commented:
also add test verifying transaction behavior

peterthomassen commented:
@nils-wisiol I addressed all action items brought up here so far. Can you please review? Thanks

peterthomassen commented:
Added commit 209f70ed05643798ba26086cbbd1128be7046277, changing serializer behavior (see issues #16, #25) and enforcing an error when a user tries to change a domain name.

Questions:
- Should [this line](https://github.com/desec-io/desec-stack/pull/36/commits/209f70ed05643798ba26086cbbd1128be7046277#diff-d93a995458c0f3a31fb831c4f43b5d27R159) somehow be moved into the model instance constructor, so that we always store a copy of the unmodified instance in `self._orig`? That way, accessing the original instance (like in `clean_fields()`) is easier, and we save a database lookup.
- When registering a user, a JSON representation with the `email` and `id` fields is returned. I cannot find where this is specified (the user model has many more fields, and we have not written our own serializer). I'll need to answer this question in order to remove the `id` field from the response.

nils-wisiol commented on api/desecapi/pdns.py:
```diff
@@ -100,14 +104,14 @@ def create_zone(name, kind='NATIVE'):
         ]
     }
-    _pdns_post('/zones', payload)
+    return _pdns_post('/zones', payload)
```
do not expose internal objects, i.e. don't return the response object

nils-wisiol commented on api/desecapi/pdns.py:
```diff
@@ -100,14 +104,14 @@ def create_zone(name, kind='NATIVE'):
     Commands pdns to delete a zone with the given name.
     """
-    _pdns_delete('/zones/' + normalize_hostname(name))
+    return _pdns_delete('/zones/' + normalize_hostname(name))
```
do not expose internal objects, i.e. don't return the response object

nils-wisiol commented on api/desecapi/pdns.py:
```diff
@@ -131,7 +135,7 @@ def set_dyn_records(name, a, aaaa):
     name = normalize_hostname(name)
 
-    _pdns_patch('/zones/' + name, {
+    return _pdns_patch('/zones/' + name, {
```
do not expose internal objects, i.e. don't return the response object

nils-wisiol commented on api/desecapi/pdns.py:
```diff
@@ -147,7 +151,7 @@ def set_rrset(zone, name, type, value):
     name = normalize_hostname(name)
 
-    _pdns_patch('/zones/' + zone, {
+    return _pdns_patch('/zones/' + zone, {
```
do not expose internal objects, i.e. don't return the response object

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -132,25 +131,33 @@ def pdns_sync(self):
+    @transaction.atomic
     def save(self, *args, **kwargs):
+        try:
+            self._orig = Domain.objects.get(pk=self.pk)
```
use `__setattr__` to track changes https://stackoverflow.com/questions/2898711/django-model-fields-getter-setter

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -62,7 +62,15 @@ def perform_create(self, serializer):
+            obj = serializer.save(owner=self.request.user)
+        except Exception as e:
+            if str(e).endswith(' already exists"}'):
+                ex = ValidationError(detail={"detail": "This domain name is unavailable.", "code": "domain-unavailable"})
```
use the same error message for 'domain-taken' above

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -99,6 +100,12 @@ class Domain(models.Model):
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
 
+    def clean_fields(self, exclude=None):
+        if self._orig is not None and self.name != self._orig.name:
```
needs to be updated, see comment above

peterthomassen commented:
I addressed your suggestions, and things are pretty much ready. The last remaining question is question 2 from https://github.com/desec-io/desec-stack/pull/36#issuecomment-281702688

peterthomassen commented:
The last remaining question was also addressed (the answer was `djoser.serializers`). Merge?
